samza-api/src/main/java/org/apache/samza/operators/descriptors/base/system/SystemDescriptor.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.descriptors.base.system;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.samza.SamzaException;
+import org.apache.samza.operators.descriptors.base.stream.OutputDescriptor;
+import org.apache.samza.serializers.Serde;
+import org.apache.samza.system.SystemStreamMetadata.OffsetType;
+
+/**
+ * The base descriptor for a system. Allows setting properties that are common to all systems.
+ *
+ * @param <SystemMessageType> default type of messages in this system.
+ * @param <SubClass> type of the concrete sub-class
+ */
+@SuppressWarnings("unchecked")
+public abstract class SystemDescriptor<SystemMessageType, SubClass extends SystemDescriptor<SystemMessageType, SubClass>> {
```

Comment:
Would you be open to using complex type parameterization in order to consolidate the "getInputDescriptor" methods into this class as abstract methods? Then the Simple/Transforming/ExpandingSystemDescriptor classes would be much simpler.

Reply:
I'm thinking about this a little more. Even type parameterization might not help.